sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs

```diff
@@ -63,5 +70,65 @@ public TClient GetClient(IServiceProvider serviceProvider, object options, Token
                 return _cachedClient;
             }
         }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_clientInitialized)
+            {
+                if (_asyncDisposable)
+                {
+                    IAsyncDisposable disposableClient;
+
+                    lock (_cacheLock)
+                    {
+                       if (!_clientInitialized)
+                       {
+                           return;
+                       }
+
+                       disposableClient = (IAsyncDisposable)_cachedClient;
+
+                       _cachedClient = default;
+                       _clientInitialized = false;
+                    }
+
+                    await disposableClient.DisposeAsync().ConfigureAwait(false);
+                }
+                else if (_disposable)
+                {
+                    Dispose();
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_clientInitialized)
+            {
+                if (_disposable)
```

Comment:
How could Dispose be called if the registration is not IDisposable?

Reply:
The registration is always disposable, but the `TClient` that it holds may not be.

Reply:
discussed offline